api/raftlease/client.go

```diff
@@ -0,0 +1,609 @@
+	// next time around we should reap them.
+	for id, remote := range c.servers {
+		if witnessed.Contains(id) {
+			continue
+		}
+
+		remote.Kill()
+
+		if err := remote.Wait(); err != nil {
+			// We don't care in reality about the death rattle of a server, as
+			// it's already dead to us.
+			c.config.Logger.Errorf("error waiting for remote server death: %v", err)
+		}
+		// Ensure we still delete the id from the server list, even though the
+		// remote Wait might have failed.
+		delete(c.servers, id)
+	}
+	return nil
+}
+
+// RemoteConfig defines the configuration for creating a NewRemote.
+type RemoteConfig struct {
+	APIInfo *api.Info
+	Clock   clock.Clock
+	Logger  Logger
+}
+
+// NewRemote creates a new Remote from a given address.
+func NewRemote(config RemoteConfig) Remote {
+	r := &remote{
+		config: config,
+	}
+	r.tomb.Go(r.loop)
+	return r
+}
+
+// RaftLeaseApplier defines a client for applying leases.
+type RaftLeaseApplier interface {
+	ApplyLease(command string) error
+}
+
+type remote struct {
+	config         RemoteConfig
+	mutex          sync.Mutex
+	tomb           tomb.Tomb
+	stopConnecting chan struct{}
+
+	api    base.APICallCloser
+	client RaftLeaseApplier
+}
+
+// Address returns the current remote server address.
+func (r *remote) Address() string {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if len(r.config.APIInfo.Addrs) == 0 {
+		return ""
+	}
+	return r.config.APIInfo.Addrs[0]
+}
+
+// SetAddress updates the current remote server address. This will cause
+// the closing of the underlying connection.
+func (r *remote) SetAddress(addr string) {
+	// They're the same address, nothing to do here.
+	if r.Address() == addr {
+		return
+	}
+
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.api == nil && r.stopConnecting != nil {
+		close(r.stopConnecting)
+		r.stopConnecting = nil
+	}
+	r.config.APIInfo.Addrs = []string{addr}
+}
+
+// Request performs a request against a specific api.
+func (r *remote) Request(ctx context.Context, command *raftlease.Command) error {
+	if r.client == nil {
+		r.config.Logger.Errorf("No client dropping command: %v", command)
+		return lease.ErrDropped
+	}
+
+	bytes, err := command.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check that the context hasn't been canceled before applying the lease.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	return r.client.ApplyLease(string(bytes))
+}
+
+// Kill is part of the worker.Worker interface.
+func (r *remote) Kill() {
+	r.mutex.Lock()
+	if r.api != nil {
+		_ = r.api.Close()
+	}
+	r.mutex.Unlock()
+	r.tomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (r *remote) Wait() error {
+	return r.tomb.Wait()
+}
+
+func (r *remote) loop() error {
+	var delay <-chan time.Time
+	for {
+		if r.api == nil {
+			if r.connect() {
+				delay = nil
+			} else {
+				// Apply a back-off if we where unable to connect.
+				delay = r.config.Clock.After(time.Second * 5)
```

Comment:
If you want something fancier for backoffs: https://github.com/juju/retry/blob/master/retry.go#L239
(Not sure if this one is pulled into the 2.9 branch though)